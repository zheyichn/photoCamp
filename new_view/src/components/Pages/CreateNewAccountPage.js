import React from "react";
import {
  Grid,
  Box,
  TextField,
  Typography,
  Button,
  Divider,
  Link,
  Alert,
} from "@mui/material";
import { ThemeProvider } from "@mui/material/styles";
import ThemeBuilder from "../../themes/ThemeBuilder";
import { useState } from "react";
import Header from "../Activity/Header";
// import { styled } from '@mui/system';

import { createUser } from "../../api/mock_api_j";

export default function CreateNewAccountPage() {
  // states to capture the account creation process
  const [success, setSuccess] = useState(false);
  const [failure, setFailure] = useState(false);
  const [err, setErr] = useState("");
  const [userName, setUsername] = useState("");

  // custom text field built on top of standard text field
  const NewAccountTextField = (props) => {
    const myProp = props;
    const label = myProp.label;
    return (
      <TextField
        id={props.id}
        label={label}
        name={label}
        variant="outlined"
        size="small"
        onChange={handleOnChange}
        sx={{
          input: { background: "#FFFFFF" },
          width: "200px",
        }}
      />
    );
  };

  // handles user inputs
  let username;
  let password;
  let email;

  const handleOnChange = (e) => {
    if (e.target.name === "Username:") {
      username = e.target.value;
    }
    if (e.target.name === "Password:") {
      password = e.target.value;
    }
    if (e.target.name === "Email:") {
      email = e.target.value;
    }
  };

  // handles creation of new account
  const handleCreateNewAccount = async (e) => {
    e.preventDefault();
    const newUser = { userName: username, password: password, email: email };
    setFailure(false);
    try {
      const newUserCreated = await createUser(newUser);
      const userName = newUserCreated.userName;
      setUsername(userName);
      setSuccess(true);
      setTimeout(() => {
        setSuccess(false);
        setUsername(userName);
        window.location.pathname = "/";
      }, 2000);
    } catch (err) {
      setErr(err.message);
      setFailure(true);
    }
  };

  return (
    <div>
      <ThemeProvider theme={ThemeBuilder()}>
        <Header />
        <Grid
          container
          spacing={0}
          direction="column"
          alignItems="center"
          justifyContent="center"
          style={{
            minHeight: "100vh",
            gap: "20px",
          }}
        >
          <Grid
            item
            xs={6}
            backgroundColor="#D9D9D9"
            borderRadius="12px"
            style={{
              minHeight: "350px",
              minWidth: "300px",
            }}
          >
            <Box
              m={2}
              sx={{
                display: "flex",
                flexDirection: "column",
                justifyContent: "center",
                alignContent: "center",
                gap: "15px",
              }}
            >
              <Typography align="center" style={{ fontSize: 24 }}>
                Create New Account
              </Typography>
              <NewAccountTextField label="Email:" id="regEmail" />
              <NewAccountTextField label="Username:" id="regUsername" />
              <NewAccountTextField label="Password:" id="regPassword" />
              {failure && <Alert severity="error">{err}</Alert>}
              {success && (
                <Alert severity="success">
                  Account created for {userName}
                  <br />
                  <b>Redirecting to login page...</b>
                </Alert>
              )}
              <Button
                variant="contained"
                size="small"
                sx={{ width: "140px" }}
                onClick={handleCreateNewAccount}
                id="signupBtn"
              >
                Sign Up
              </Button>
              <Divider sx={{ backgroundColor: "#000000" }} />
              <Typography style={{ fontSize: 12 }}>
                <Link href="/">I already have an account</Link>
              </Typography>
            </Box>
          </Grid>
          <Box sx={{ height: 100 }}></Box>
        </Grid>
      </ThemeProvider>
    </div>
  );
}
