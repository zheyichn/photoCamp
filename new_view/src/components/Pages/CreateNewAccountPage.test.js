/**
 * @jest-environment jsdom
 */
import "@testing-library/jest-dom/extend-expect";
import { render, screen, cleanup, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";
import CreateNewAccountPage from "./CreateNewAccountPage";

// afterEach function runs after each test suite is executed
afterEach(() => {
  cleanup();
});

describe("Create New Account Page", () => {
  // Test1: match snapshots
  it("Page renders correctly", () => {
    const tree = renderer.create(<CreateNewAccountPage />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  // Test2: check create new account textboxes
  it("Create New Account textboxes handle input correctly", async () => {
    render(<CreateNewAccountPage />);
    const element = screen.getAllByRole("textbox");
    await userEvent.type(element[0], "this is a email");
    await userEvent.type(element[1], "this is a username");
    await userEvent.type(element[2], "this is a password");

    // assertion: verify that the text is in the textbox
    expect(element[0]).toHaveValue("this is a email");
    expect(element[1]).toHaveValue("this is a username");
    expect(element[2]).toHaveValue("this is a password");
  });

  // Test3: Check if alert is fired when button is clicked
  // it("Check if alert is thrown when button is clicked", async() => {
  //   render(<CreateNewAccountPage />);
  //   await userEvent.click(screen.getByRole('button'));
  // })
});
