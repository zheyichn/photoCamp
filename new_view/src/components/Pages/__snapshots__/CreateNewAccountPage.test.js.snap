// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create New Account Page Page renders correctly 1`] = `
<div>
  <div>
    <div
      className="MuiGrid-root css-b5hrgx-MuiGrid-root"
    >
      <div
        className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      >
        <div
          className="MuiGrid-root MuiGrid-container css-1q1yn3m-MuiGrid-root"
          style={
            {
              "backgroundColor": "white",
            }
          }
        >
          <div
            className="MuiGrid-root MuiGrid-item css-l2wqce-MuiGrid-root"
            style={
              {
                "backgroundColor": "white",
              }
            }
          >
            <div
              className="MuiTypography-root MuiTypography-h6 css-1qvm0tf-MuiTypography-root"
            >
              PhotoCamp
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-fdtuls-MuiGrid-root"
    style={
      {
        "gap": "20px",
        "minHeight": "100vh",
      }
    }
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-vu25ta-MuiGrid-root"
      style={
        {
          "minHeight": "350px",
          "minWidth": "300px",
        }
      }
    >
      <div
        className="MuiBox-root css-5cp77"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter css-3mddyt-MuiTypography-root"
          style={
            {
              "fontSize": 24,
            }
          }
        >
          Create New Account
        </p>
        <div
          className="MuiFormControl-root MuiTextField-root css-1c01b1h-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-13tblej-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor=":r0:"
            id=":r0:-label"
          >
            Email:
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-175sm4b-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id=":r0:"
              name="Email:"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-yjsfm1"
              >
                <span>
                  Email:
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          className="MuiFormControl-root MuiTextField-root css-1c01b1h-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-13tblej-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor=":r1:"
            id=":r1:-label"
          >
            Username:
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-175sm4b-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id=":r1:"
              name="Username:"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-yjsfm1"
              >
                <span>
                  Username:
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          className="MuiFormControl-root MuiTextField-root css-1c01b1h-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-13tblej-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor=":r2:"
            id=":r2:-label"
          >
            Password:
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-175sm4b-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id=":r2:"
              name="Password:"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-yjsfm1"
              >
                <span>
                  Password:
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall css-1lk5wtx-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Sign Up
        </button>
        <hr
          className="MuiDivider-root MuiDivider-fullWidth css-ogpflq-MuiDivider-root"
        />
        <p
          className="MuiTypography-root MuiTypography-body1 css-uckjdc-MuiTypography-root"
          style={
            {
              "fontSize": 12,
            }
          }
        >
          <a
            className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
            href="/"
            onBlur={[Function]}
            onFocus={[Function]}
          >
            I already have an account
          </a>
        </p>
      </div>
    </div>
    <div
      className="MuiBox-root css-73nay0"
    />
  </div>
</div>
`;
